ProofWidgets/Component/Panel/SelectInsertPanel.lean

```diff
@@ -0,0 +1,113 @@
+      let some firstLoc := selectedLocations[0]? | return none
+      let doc ← RequestM.readDoc
+      ctx.runMetaM {} do
+        let md ← firstLoc.mvarId.getDecl
+        let lctx := md.lctx |>.sanitizeNames.run' {options := (← getOptions)}
+        Meta.withLCtx lctx md.localInstances do
+          let insertedCode ← mkCmdStr selectedLocations md.type
+          let textDocumentEdit : Lsp.TextDocumentEdit := {
+            textDocument := { uri := doc.meta.uri, version? := doc.meta.version },
+            edits        := #[{range := replaceRange, newText := insertedCode }] }
+          return some { content := insertedCode,
+                        edit := Lsp.WorkspaceEdit.ofTextDocumentEdit textDocumentEdit,
+                        /- CHECKME: the new cursor position computation below looks very naive.
+                           Will it work with non-ascii insertedCode? -/
+                        newCursorPos := {line := replaceRange.start.line,
+                                         character := replaceRange.start.character + insertedCode.length } }
+
+/-- Props that will be passed to the JavaScript component on top of the props inserted
+automatically by `savePanelWidgetInfo`. -/
+structure InsertPanelProps where
+  /-- The panel title. -/
+  title : String
+  /-- The help text displayed when nothing is selected. -/
+  help : String
+  /-- The name of the callback function as a String. -/
+  callback : String
+  /-- The range to be replaced. -/
+  replaceRange : Lsp.Range
+  deriving FromJson, ToJson
+
+@[widget_module]
+def mkInsertWidget : ProofWidgets.Component InsertPanelProps where
+  javascript := include_str ".."/".."/".."/"build"/"js"/"selectInsertPanel.js"
+
+open Elab Tactic ProofWidgets in
+/-- Create the tactic that will display the widget. -/
+def mkInsertTactic (panel: Name) (title help : String) (callback : Name) (stx : Syntax) :
+    TacticM Unit := do
+  let some replaceRange := (← getFileMap).rangeOfStx? stx | return
+  let props : InsertPanelProps := {
+    title := title
+    help := help
+    callback := callback.toString
+    replaceRange := replaceRange
+  }
+  savePanelWidgetInfo stx panel (pure <| toJson props)
+
+/-- Define a tactic with the given name that create a panels with the given title
+and help text that allow to select some locations in the tactic state and then call
+the given callback function to generate a String that will replace the tactic when
+users click on a link.  The callback function must have type
+`(pos : Array Lean.SubExpr.GoalsLocation) → (goalType : Expr) → MetaM String`. -/
+macro "mkSelectInsertTactic" name:str title:str help:str callback:ident : command =>
+  let cmdName : Name := .str `mkSelectInsertCmd (name.getString ++ "_widget")
+  `(
+    @[server_rpc_method]
+    def $(mkIdent cmdName) := mkSelectInsertCommand $callback
+    elab stx:$name:str : $(mkIdent `tactic) =>
+      mkInsertTactic ``mkInsertWidget $title:str $help:str $(quote cmdName) stx)
```

Comment:
I am not a fan of macro-generating macros like this. I am wary of ending up in a situation where our code cannot be read without reading a bunch of macro implementations first, thus adding a prerequisite of 'user knows metaprogramming'. I would prefer to stay closer to object-level whenever possible, even if that means writing a bit more boilerplate every time: I mean something like for example `SelectInsertConv2.lean` where I am tagging the RPC method by hand.

Reply:
On the other hand one could argue that just a `mkSelectInsertTactic` is more readable if one is willing to treat it as a black box! But in that case we should make it a bit more general so as to not have a bunch of `mkDoFooTactic` macros for every `Foo` that one might want to do.